### PR TITLE
feat(plugin): run tests with mux server

### PR DESCRIPTION
**Pull request by byashimov**
## About this change—what it does

Runs tests with mux server involving both protocol 5 and 6.
That makes possible to test resources from both: sdk and plugin framework.

### CODE REVIEW

**Review by Serpentiel — CHANGES_REQUESTED** [review]
we already have `Protov6ProviderFactories` generated in the current version of the code: https://github.com/aiven/terraform-provider-aiven/blob/main/internal/provider/provider_test.go#L19-L25

why don't we simply make variable at line 23 public, and then it will be possible to use it in the tests that are being ran from the SDK provider?

I don't understand why do we need this massive refactoring, when we already have `ProtoV6ProviderFactories` in the current version

File: internal/server/server.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	server, err := tf6muxserver.NewMuxServer(ctx, providers...)
+	if err != nil {
+		return nil, err
+	}
+	return server.ProviderServer(), nil
+}
+
+func NewTestProviderServerFactories() map[string]func() (tfprotov6.ProviderServer, error) {
+	return map[string]func() (tfprotov6.ProviderServer, error){
+		Name: func() (tfprotov6.ProviderServer, error) {
```
Comment (Serpentiel) [inline internal/server/server.go:43]: this should be called `"aiven"`, not `"registry.terraform.io/aiven/aiven"`

**Review by Serpentiel — CHANGES_REQUESTED** [review]

**Comment by byashimov** [comment]
1. [tests](https://github.com/aiven/terraform-provider-aiven/actions/runs/5996656971)
2. [mixed test](https://github.com/aiven/terraform-provider-aiven/actions/runs/5996656971/job/16263108969#step:6:15)

File: internal/server/server.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package server
```
Comment (Serpentiel) [inline internal/server/server.go:1]: nit: comment please

File: internal/server/server.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/hashicorp/terraform-plugin-framework/providerserver"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-mux/tf5to6server"
+	"github.com/hashicorp/terraform-plugin-mux/tf6muxserver"
+
+	"github.com/aiven/terraform-provider-aiven/internal/plugin"
+	sdk "github.com/aiven/terraform-provider-aiven/internal/sdkprovider/provider"
+)
+
+func NewMuxServer(ctx context.Context, version string) (tfprotov6.ProviderServer, error) {
```
Comment (Serpentiel) [inline internal/server/server.go:15]: nit: comment please

File: internal/server/server.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	sdkProvider, err := tf5to6server.UpgradeServer(
+		ctx,
+		sdk.Provider(version).GRPCProvider,
+	)
+
```
Comment (Serpentiel) [inline internal/server/server.go:20]: nit: remove NL

File: internal/server/server.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+			return sdkProvider
+		},
+		providerserver.NewProtocol6(plugin.New(version)()),
+	}
+
+	server, err := tf6muxserver.NewMuxServer(ctx, providers...)
+	if err != nil {
+		return nil, err
+	}
+	return server.ProviderServer(), nil
```
Comment (Serpentiel) [inline internal/server/server.go:36]: nit: NL here

File: main.go [hunk 6]
```diff
@@ -24,37 +20,26 @@ func main() {
...
+	err = tf6server.Serve(
 		"registry.terraform.io/aiven/aiven",
-		muxServer.ProviderServer,
+		func() tfprotov6.ProviderServer {
+			return muxServer
+		},
 		serveOpts...,
-	); err != nil {
+	)
+
```
Comment (Serpentiel) [inline main.go:41]: nit: remove NL

File: internal/acctest/acctest.go [hunk 7]
```diff
@@ -1,49 +1,67 @@
...
+
+func GetTestAivenClient() *aiven.Client {
+	testAivenClientOnce.Do(func() {
+		client, err := common.NewAivenClient()
+		if err != nil {
+			log.Fatal(err)
+		}
+		testAivenClient = client
+	})
+	return testAivenClient
```
Comment (Serpentiel) [inline internal/acctest/acctest.go:40]: nit: remove NL

Comment (ivan-savciuc) [reply internal/acctest/acctest.go:40]: can you explain please?

**Review by ivan-savciuc — APPROVED** [review]
LGTM

File: internal/acctest/acctest.go (hunk 7 above)
Comment (ivan-savciuc) [reply internal/acctest/acctest.go:40]: ah new line I guess :)

File: internal/acctest/acctest.go [hunk 8]
```diff
@@ -1,49 +1,67 @@
...
-func init() {
-	TestAccProvider = provider.Provider("test")
-	TestAccProviderFactories = map[string]func() (*schema.Provider, error){
-		"aiven": func() (*schema.Provider, error) {
-			return TestAccProvider, nil
+	testAivenClient              *aiven.Client
+	testAivenClientOnce          sync.Once
+	TestProtoV6ProviderFactories = map[string]func() (tfprotov6.ProviderServer, error){
+		"aiven": func() (tfprotov6.ProviderServer, error) {
+			return server.NewMuxServer(context.Background(), "test")
```
[the remark this answers is not in the document]
Comment (byashimov) [reply internal/acctest/acctest.go:27]: Well it is never shared. It is used exactly here only.

**Pull request merged**